### PR TITLE
[Example] add implementation of grace

### PR DESCRIPTION
**Pull request by hengruizhang98**
## Description
Add implementation of the paper 'Deep Graph Contrastive Representation Learning'.
@zhjwy9343 
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/pytorch/grace/main.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import yaml
+from yaml import SafeLoader
+
+from eval import label_classification
+import warnings
+
+warnings.filterwarnings('ignore')
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--dataname', type=str, default='cora')
```
Comment (mufeili) [inline examples/pytorch/grace/main.py:19]: Add `choices=[...]`

File: examples/pytorch/grace/main.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+parser.add_argument('--gpu', type=int, default=0)
+parser.add_argument('--split', type=str, default='random')
```
Comment (mufeili) [inline examples/pytorch/grace/main.py:21]: Add `choices=[...]`

File: examples/pytorch/grace/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+```
+--dataname         str     The graph dataset name.                Default is 'cora'.
+--gpu              int     GPU index.                             Default is 0.
+--split            int     Dataset spliting method.               Default is 'random'.
+```
+
+## How to run examples
+
+In the paper(as well as authors' repo), the training set and testing set are split randomly with 1:9 ratio. In order to fairly compare it with other models with public split, in this repo we also provide its results using public split. To run the examples, follow the following instructions.
```
Comment (mufeili) [inline examples/pytorch/grace/README.md:39]: Define "public split" the first time. For example, "with public split" -> "with the split they used, termed public split,"

Comment (mufeili) [inline examples/pytorch/grace/README.md:39]: "its results" -> "experiment results"

Comment (mufeili) [inline examples/pytorch/grace/README.md:39]: "follow the following" -> "use the following"

File: examples/pytorch/grace/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+```python
+# Cora with random split
+python main.py --dataname cora
+
+# Cora with public split
+python main.py --dataname cora --split public
+```
+
+replace 'cora' with 'citeseer' or 'pubmed' if you would like to run this example on other datasets.
```
Comment (mufeili) [inline examples/pytorch/grace/README.md:49]: "on" -> "for"

File: examples/pytorch/grace/README.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+## 	Performance
+
+We use the same hyper-parameter settings as provided by the author, you can check config.yaml for detailed hyper-parameters for each dataset.
```
Comment (mufeili) [inline examples/pytorch/grace/README.md:53]: "hyper-parameter" -> "hyperparameter"

File: examples/pytorch/grace/config.yaml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  activation: 'relu'
+  num_layers: 2
+  drop_edge_rate_1: 0.4
+  drop_edge_rate_2: 0.1
+  drop_feature_rate_1: 0.0
+  drop_feature_rate_2: 0.2
+  tau: 0.7
+  num_epochs: 1500
+  weight_decay: 0.00001
+dblp:
```
Comment (mufeili) [inline examples/pytorch/grace/config.yaml:40]: Remove the part for dblp.

File: examples/pytorch/grace/main.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+if args.gpu != -1 and th.cuda.is_available():
+    args.device = 'cuda:{}'.format(args.gpu)
+else:
+    args.device = 'cpu'
+
+
+if __name__ == '__main__':
+
+    # Step 1: Load hyper-parameters =================================================================== #
```
Comment (mufeili) [inline examples/pytorch/grace/main.py:32]: "hyper-parameters " -> "hyperparameters"

File: examples/pytorch/grace/dataset.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    train_mask = graph.ndata.pop('train_mask')
+    val_mask = graph.ndata.pop('val_mask')
+    test_mask = graph.ndata.pop('test_mask')
+
+    num_class = dataset.num_classes
+    feat = graph.ndata.pop('feat')
+    labels = graph.ndata.pop('label')
+
+    return graph, feat, labels, num_class, train_mask, val_mask, test_mask
```
Comment (mufeili) [inline examples/pytorch/grace/dataset.py:21]: `num_class` and `val_mask` will not be used.

File: examples/pytorch/grace/model.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+class GCN(nn.Module):
+    def __init__(self, in_dim, out_dim, act_fn, num_layers = 2):
+        super(GCN, self).__init__()
+
+        assert num_layers >= 2
+        self.num_layers = num_layers
+        self.convs = nn.ModuleList()
+
+        self.convs.append(GraphConv(in_dim, out_dim * 2))
+        for _ in range(self.num_layers - 2):
```
Comment (mufeili) [inline examples/pytorch/grace/model.py:17]: This does not come into effect at all. I guess we can remove the argument and the related code.

File: examples/pytorch/grace/model.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import torch.nn.functional as F
+from dgl.nn import GraphConv
+
+
+# Multi-layer Graph Convolutional Networks
+class GCN(nn.Module):
+    def __init__(self, in_dim, out_dim, act_fn, num_layers = 2):
+        super(GCN, self).__init__()
+
+        assert num_layers >= 2
```
Comment (mufeili) [inline examples/pytorch/grace/model.py:12]: Does not come into effect

File: examples/pytorch/grace/model.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+class Grace(nn.Module):
+    r"""
+        GRACE model
+    Parameters
+    -----------
+    in_dim: int
+        Input feature size.
+    hid_dim: int
+        Hidden feature size
```
Comment (mufeili) [inline examples/pytorch/grace/model.py:49]: "Hidden feature size" -> "Hidden feature size."

File: examples/pytorch/grace/model.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    out_dim: int
+        Output feature size.
+    num_layers: int
+        Number of the GNN encoder layers.
+    act_fn: nn.Module
+        Activation function
```
Comment (mufeili) [inline examples/pytorch/grace/model.py:55]: "Activation function" -> "Activation function."

File: examples/pytorch/grace/model.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    temp: float
+        temperature constant
```
Comment (mufeili) [inline examples/pytorch/grace/model.py:57]: "temperature constant" -> "Temperature constant."

File: examples/pytorch/grace/aug.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+def aug(graph, x, feat_drop_rate, edge_mask_rate):
+    ng = drop_edge(graph, edge_mask_rate)
+    feat = drop_feat(x, feat_drop_rate)
+    ng = ng.add_self_loop()
+
+    return ng, feat
+
+def drop_edge(graph, drop_prob):
+    N = graph.number_of_nodes()
```
Comment (mufeili) [inline examples/pytorch/grace/aug.py:15]: "number_of_nodes" -> "num_nodes"

File: examples/pytorch/grace/aug.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    E = graph.number_of_edges()
```
Comment (mufeili) [inline examples/pytorch/grace/aug.py:16]: "number_of_edges" -> "num_edges"

File: examples/pytorch/grace/model.py (hunk 9 above)
Comment (hengruizhang98) [reply examples/pytorch/grace/model.py:17]: I guess users might try a different number of layers, so we can just keep it.

**Comment by mufeili** [comment]
Done a first pass.

File: examples/pytorch/grace/model.py (hunk 9 above)
Comment (mufeili) [reply examples/pytorch/grace/model.py:17]: Fine

**Review by mufeili — APPROVED** [review]
LGTM

**Pull request merged**